kevm-pyk/src/kontrol/foundry.py

```diff
@@ -758,37 +800,87 @@ def _init_and_run_proof(_init_problem: tuple[str, str, str | None]) -> tuple[boo
-        else:
-            _apr_proofs = []
-            for init_problem in init_problems:
-                _apr_proofs.append(_init_and_run_proof(init_problem))
+            return proof
+
+    def run_cfg_group(tests: list[tuple[str, str | None]]) -> list[APRProof]:
+        llvm_definition_dir = foundry.llvm_library if use_booster else None
+
+        def create_server() -> KoreServer:
+            return kore_server(
+                definition_dir=foundry.kevm.definition_dir,
+                llvm_definition_dir=llvm_definition_dir,
+                module_name=foundry.kevm.main_module,
+                command=kore_rpc_command,
+                bug_report=bug_report,
+                smt_timeout=smt_timeout,
+                smt_retry_limit=smt_retry_limit,
+            )
+
+        class Task:
+            ...
+
+        class CreateAndRunTask(Task):
+            _test: str
+            _id: str | None
+
+            def __init__(self, test: str, id: str | None):
+                self._test = test
+                self._id = id
+
+        class RunTask(Task):
+            _proof: Proof
+            _port: int
+
+            def __init__(self, proof: Proof, port: int):
+                self._proof = proof
+                self._port = port
+
+        results = []
+        futures: dict[str, Future] = {}
+        executor = ThreadPoolExecutor(max_workers=workers)
+
+        tasks: list[Task] = [CreateAndRunTask(test, id) for test, id in tests]
+        base_proofs = {test: test for test, _ in tests}
+        servers = {}
+
+        for test, _ in tests:
+            servers[test] = create_server()
+
+        while len(tasks) > 0 or len(futures) > 0:
+            if len(tasks) > 0:
+                task = tasks.pop()
 
-        apr_proofs = dict(zip(tests, _apr_proofs, strict=True))
-        return apr_proofs
+                if type(task) is CreateAndRunTask:
+                    contract, method = task._test.split('.')
+
+                    futures[task._test] = executor.submit(_init_and_run_proof, (contract, method, task._id))
+
+                elif type(task) is RunTask:
+                    futures[task._proof.id] = executor.submit(_run_proof, (task._port, task._proof))
+
+            if len(futures) > 0:
+                done_futures = [(_proof, future) for (_proof, future) in futures.items() if future.done()]
+                futures = {_proof: future for (_proof, future) in futures.items() if not future.done()}
+                for _proof, future in done_futures:
+                    proof = future.result()
+                    results.append(proof)
+
+                    subproofs = proof.subproofs
+                    for subproof in subproofs:
+                        base_proofs[subproof.id] = base_proofs[_proof]
+                        port = servers[base_proofs[subproof.id]].port
+                        tasks.append(RunTask(subproof, port))
+
+            wait(futures.values())
```

Comment:
If I understand correctly, at this point all futures are waited before pending tasks are submitted. If so, this is not good for utilization.

Reply:
Btw, are two lists (`futures`, `tasks`) necessary? Given a single list of `pedning` futures, in each iteration you can try something like
1. If `pending` is empty then `break`.
2. Otherwise, if there are no `done` futures in `pending` then `continue`.
3. Otherwise, get a `done` future from `pending` and check it's results.
4. If based on the results there are new tasks to perform, submit them, and add all new futures to `pending`.
5. Remove the future from `pending`.

Reply:
I didn't realize this was what `wait` was doing. I've added the parameter `return_when='FIRST_COMPLETED'`, so it should wait for any, not all now.
I've also applied your suggestion and now there is only one list, and we just immediately add tasks to it rather than needing to wait for the next iteration.